### PR TITLE
chore(weave): refactor objects query into a simple query builder

### PR DESCRIPTION
**Pull request by gtarpenning**
## Description

[WB-22419](https://wandb.atlassian.net/browse/WB-22419)

Straight refactor for readability. Use a query builder to handle adding conditions to the objects query. Considerably less extensive or flexible as the other query builders, but the objects query is also much more complex. Not sure we need as much flexibility here, this pr should be a balance between risky and merely surface level. 

## Testing

Existing code coverage + unit tests

[WB-22419]: https://wandb.atlassian.net/browse/WB-22419?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by gtarpenning** [comment]
* **#3223** <a href="https://app.graphite.dev/github/pr/wandb/weave/3223?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/wandb/weave/3223?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: tests/trace/test_objects_query_builder.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    assert _make_sort_part(sort_by) == "ORDER BY created_at ASC"
+
+    sort_by = [
+        tsi.SortBy(field="created_at", direction="desc"),
+        tsi.SortBy(field="object_id", direction="asc"),
+    ]
+    assert _make_sort_part(sort_by) == "ORDER BY created_at DESC, object_id ASC"
+
+    # Invalid sort fields should be ignored
+    sort_by = [tsi.SortBy(field="invalid_field", direction="asc")]
```
Comment (tssweeney) [inline tests/trace/test_objects_query_builder.py:40]: not error?

File: tests/trace/test_objects_query_builder.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    assert _make_conditions_part([]) == ""
+    assert _make_conditions_part(["condition1"]) == "WHERE condition1"
+    assert (
+        _make_conditions_part(["condition1", "condition2"])
+        == "WHERE ((condition1) AND (condition2))"
+    )
+
+
+def test_make_object_id_conditions_part():
+    assert _make_object_id_conditions_part(None) == ""
```
Comment (tssweeney) [inline tests/trace/test_objects_query_builder.py:55]: seems like a weird function to me - it assumes that it always comes after a non-empty where clause?

File: tests/trace/test_objects_query_builder.py (hunk 1 above)
Comment (gtarpenning) [reply tests/trace/test_objects_query_builder.py:40]: hmmm, yeah lets error

File: tests/trace/test_objects_query_builder.py (hunk 2 above)
Comment (gtarpenning) [reply tests/trace/test_objects_query_builder.py:55]: Its meant to be an internal fn but i see your point. i'll rename and add a bit of logic here

File: tests/trace/test_objects_query_builder.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    ObjectQueryBuilder,
+    _make_conditions_part,
+    _make_limit_part,
+    _make_object_id_conditions_part,
+    _make_offset_part,
+    _make_sort_part,
+)
+
+
+def test_make_limit_part():
```
Comment (tssweeney) [inline tests/trace/test_objects_query_builder.py:14]: I think a lot of this test suite would be improved if you were to test the public api exposed by your query build as opposed to the private helper functions used internal to construct the final results

Comment (gtarpenning) [reply tests/trace/test_objects_query_builder.py:14]: - unit tests [in this pr]
- end to end tests [exist already]
- object_query_builder public method tests [todo] 

got it

Comment (gtarpenning) [reply tests/trace/test_objects_query_builder.py:14]: will add those

Comment (tssweeney) [reply tests/trace/test_objects_query_builder.py:14]: the other one is just testing the final constructed sql... i don't think you have any tests that demonstrate the query builders primary function

File: weave/trace_server/objects_query_builder.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+            )
+            {self.conditions_part}
+            {self.sort_part}
+            {self.limit_part}
+            {self.offset_part}
+        """
+
+
+def make_objects_val_query_and_parameters(
+    project_id: str, object_ids: list[str], digests: list[str]
```
Comment (tssweeney) [inline weave/trace_server/objects_query_builder.py:260]: i would add a test for this one

File: weave/trace_server/objects_query_builder.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        if offset < 0:
+            raise ValueError("Offset must be a positive integer")
+        if self._offset is not None:
+            raise ValueError("Offset can only be set once")
+        self._offset = offset
+
+    def set_metadata_only(self, metadata_only: bool) -> None:
+        self.metadata_only = metadata_only
+
+    def make_metadata_query(self) -> str:
```
Comment (tssweeney) [inline weave/trace_server/objects_query_builder.py:208]: i would add a test for this one

File: weave/trace_server/objects_query_builder.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        # Add an empty val_dump to the end of the row
+        row_with_val_dump = row + ("{}",)
+        columns_with_val_dump = OBJECT_METADATA_COLUMNS + ["val_dump"]
+        row_dict = dict(zip(columns_with_val_dump, row_with_val_dump))
+        row_model = SelectableCHObjSchema.model_validate(row_dict)
+        result.append(row_model)
+    return result
+
+
+class ObjectQueryBuilder:
```
Comment (tssweeney) [inline weave/trace_server/objects_query_builder.py:88]: as authored, this is a "MetadataQueryBuilder" since you can't get the vals out

Comment (gtarpenning) [reply weave/trace_server/objects_query_builder.py:88]: Hmm true. I'll add the value query...

Comment (gtarpenning) [reply weave/trace_server/objects_query_builder.py]: ended up just renaming, the value query is so simple it doesn't really make sense to refactor

File: weave/trace_server/objects_query_builder.py (hunk 5 above)
Comment (gtarpenning) [reply weave/trace_server/objects_query_builder.py:198]: done

File: weave/trace_server/objects_query_builder.py (hunk 4 above)
Comment (gtarpenning) [reply weave/trace_server/objects_query_builder.py:255]: this is a funny test but done

File: tests/trace/test_objects_query_builder.py (hunk 1 above)
Comment (gtarpenning) [reply tests/trace/test_objects_query_builder.py:41]: This is really just testing the internal method, the way to add a sort by field is with `builder.add_order` and the impl checks correctness and throws there

**Review by tssweeney — APPROVED** [review]

**Pull request merged**